make the task cleanup wait time configurable

gregwebs opened the pull request:
resolves #70

I have not yet tested this in any manual or automated way. How can I start running this forked version on my ECS cluster?

gregwebs commented:
I figured out how to test this locally, and it works. I am using `1s` and I see an immediate job cleanup. However, I don't know how to look at the test case failures (very new go user here).

samuelkarp commented:
Hi @gregwebs,

Thanks for the contribution!  Before we can merge this, I'd like to see if you can address the following concerns:

1. It looks like the existing unit tests are now failing (see the [Travis CI](https://travis-ci.org/aws/amazon-ecs-agent/builds/103230162) build output).  My guess as to the cause would be [`task.engine.cfg.CleanupWaitDuration` ](https://github.com/KariusDx/amazon-ecs-agent/blob/87e79a8/agent/engine/task_manager.go#L163) now has a different value than expected in the test case (it may be uninitialized while the test case expects `taskStoppedDuration` of 3 hours).
2. I think the format you have in `README` isn't quite correct; I don't see code that handles a literal `3 hours` as the value.  Can you switch the example column to show input that would be parsed correctly and add a value for the `Default Value` column as well?
3. If possible, it'd be great to add a new unit test around the parsing code you've extracted out in `config.go`.  It'd be especially good to have a test exercising the new `ECS_ENGINE_CLEANUP_WAIT_DURATION` environment variable in terms of how it would be expected that someone would set it.
4. I can see that you're calling `DefaultConfig()` multiple times in order to get default values.  You should be able to safely let those be `0` in the `EnvironmentConfig` method, as `Merge` should properly handle the defaults.
5. Can you run `gofmt` over the code?  There's a few formatting things that would be fixed by that.

If you're unable to take care of these concerns, we can potentially take what you've written and handle them when we have the time to come back to this.

I'd also like you to please confirm this contribution is under the terms of the Apache 2.0 license.

Thanks!
Sam

gregwebs commented:
Confirming Apache 2.0 license.

I will see if I can meet you half way on cleaning this up. Thanks for the tips!

```
I can see that you're calling DefaultConfig() multiple times in order to get default values. You should be able to safely let those be 0 in the EnvironmentConfig method, as Merge should properly handle the defaults.
```

That makes sense, but now I have to understand a lot of source code to understand what `0` means. The changes I made make it very clear what is going on without having to understand a merge method. Using a constant could be a middle ground, but then the change is pretty much that already. Returning an optional type would make more sense to me.

samuelkarp commented:
Yes, I think an optional type would be great here.  Unfortunately, the closest that you can get with that in Go is a pointer where `nil` is equivalent to absent, but that requires `nil`-checking every time that the value is being read.  The zero-valued approach attempts to compromise by making the [zero value](https://tour.golang.org/basics/12) equivalent to absent, and then using [reflection](https://github.com/aws/amazon-ecs-agent/blob/v1.7.1/agent/utils/utils.go#L39-L58) to determine whether the value is the zero value during [merging](https://github.com/aws/amazon-ecs-agent/blob/v1.7.1/agent/config/config.go#L56).  The intent of all this complexity is that values can be set in `FileConfig()`, `EnvironmentConfig()`, and `EC2MetadataConfig()` and all merge correctly; setting things to the values used in `DefaultConfig()` breaks the zero value detection and thus breaks the contract.

It'd probably be clearer to just leave the values unset (which ends up being the zero value).  For example:

```go
func parseEnvVariableInt(envVar string) uint16 {
	var var16 uint16
	envVal := os.Getenv(envVar)
	if envVal == "" {
		// var16 is unset, so it has the zero-value of 0
		return var16
	}
	...
}
```

An alternate approach would be an internal `config` struct where each field is a pointer, then `Merge()` could just check for `nil` and render into the non-pointer `Config` struct at the end.  That's a larger change though, so I don't know if you want to do that in the context of your pull request.

gregwebs commented:
I think I addressed most of your feedback points. However:

* I don't know how to see the test failure. I did try googling and playing around with the commands, but I bet you know how to do this in a second :)
* `go fmt` had a ton of changes unrelated to my code: I think it is better not to put it on this PR

Feel free to make any changes you need to get this merged.
As I said, I have tested this in the wild. I can see it working both in the debug logs of ecs-agent and by the fact that stopped docker processes are being quickly removed.

samuelkarp commented:
From the stack trace in the test failure, it looks like it's failing inside `TestBatchContainerHappyPath`.  You'll note that the test creates a new `&config.Config{}` with zero-values.  Since the test is timing out (`panic: test timed out after 25s`), I think having 0 for that value does not work; fortunately `DefaultConfig()` sets it to 3 hours.  My guess would be that if you set `CleanupWaitDuration` to be 3 hours the test will likely pass (note that it tries to test behavior after 4 hours through some trickery with a time test double that I'm [part-way through replacing](https://github.com/aws/amazon-ecs-agent/pull/259)).

I have a few more line-specific comments, but I want to thank you again for the contribution here!

samuelkarp commented on README.md:
```diff
@@ -115,6 +115,7 @@ configure them as something other than the defaults.
 | `ECS_DISABLE_PRIVILEGED` | `true` | Whether launching privileged containers is disabled on the Container Instance. | `false` |
 | `ECS_SELINUX_CAPABLE` | `true` | Whether SELinux is available on the Container Instance. | `false` |
 | `ECS_APPARMOR_CAPABLE` | `true` | Whether AppArmor is available on the Container Instance. | `false` |
+| `ECS_ENGINE_CLEANUP_WAIT_DURATION` | 3h | Time duration to wait from when a task is stopped until the docker container is removed |
```
Can you add `3h` as the default value in the last column?

samuelkarp commented on agent/config/config.go:
```diff
@@ -181,6 +183,39 @@ func FileConfig() Config {
 	return config
 }
 
+func parseEnvVariableInt(envVar string, defaultValue uint16) (uint16) {
```
Nitpick: Can you move `parseEnvVariableInt` and `parseEnvVariableDuration` below `EnvironmentConfig`?  One of the things I try to do in terms of code organization is have the caller above the callee so that a reader sees the highest abstraction layer first and can go further down for details when necessary.

samuelkarp commented on agent/config/config.go:
```diff
@@ -233,20 +268,9 @@ func EnvironmentConfig() Config {
-	}
+	reservedMemory := parseEnvVariableInt("ECS_RESERVED_MEMORY", 0)
+
+	cleanupWaitDuration := parseEnvVariableDuration("ECS_ENGINE_CLEANUP_WAIT_DURATION", DefaultConfig().CleanupWaitDuration)
```
Can you either remove the default value from `parseEnvVariableDuration` (as [demonstrated](https://github.com/aws/amazon-ecs-agent/pull/286#issuecomment-172711885)) or set this to `0 * time.Millisecond`?  Having it set to a value here will cause problems with `Merge()`.

gregwebs commented:
I am busy working around ECS errors like this right now: CannotPullECRContainerError: ThrottlingException: Rate exceeded

Will let you know when I can get back to this.

samuelkarp commented:
@gregwebs If you can send me some details (account and timestamps) I can send that over to my colleagues to take a look.  skarp at amazon.com.

aaithal commented:
@gregwebs  Thank you for this PR. I will check this out locally and make some modifications to ensure that it passes tests and address @samuelkarp 's comments.

gregwebs commented:
@aaithal Thanks! We have been successfully using this in production. I have been busy making our ECS usage stable. It might finally be there now other than for the mentioned throttling that I have a support issue open for.

samuelkarp commented:
Merged with https://github.com/aws/amazon-ecs-agent/pull/302.

samuelkarp commented:
@gregwebs Thank you for this contribution!
